File: membrain-node/src/graph.ts
/**
 * Membrain graph memory layer client.
 */

import koffi from "koffi";

import type { GraphPruningResult, GraphQueryResult } from "./types";

import {
  charPtr,
  charPtrPtr,
  findLibrary,
  int64Ptr,
  MEMBRAIN_OK,
  MembrainError,
  newInt64Out,
  readInt64,
  voidPtr,
} from "./ffi";

/** Graph memory layer client. */
export class MembrainGraph {
  private handle: any;
  private lib: ReturnType<typeof koffi.load>;

  // Bound function references
  private fn_last_error: any;
  private fn_string_free: any;
  private fn_graph_free: any;
  private fn_graph_add_node: any;
  private fn_graph_remove_node: any;
  private fn_graph_query: any;
  private fn_graph_node_count: any;
  private fn_graph_edge_count: any;
  private fn_graph_prune: any;
  private fn_graph_save: any;
  private fn_graph_load: any;

  constructor(config?: Record<string, any>, libPath?: string) {
    const libFile = libPath ?? findLibrary();
    this.lib = koffi.load(libFile);
    this.bindGraphFunctions();

    if (config) {
      const createWithConfig = this.lib.func(
        "membrain_graph_new_with_config",
        voidPtr,
        ["str"]
      );
      this.handle = createWithConfig(JSON.stringify(config));
    } else {
      const create = this.lib.func("membrain_graph_new", voidPtr, []);
      this.handle = create();
    }

    if (!this.handle) {
      const err = this.getLastError();
      throw new MembrainError(`failed to create graph: ${err}`);
    }
  }

  /** Load a graph from a base64-encoded string. */
  static load(data: string, libPath?: string): MembrainGraph {
    const libFile = libPath ?? findLibrary();
    const lib = koffi.load(libFile);

    const fn_load = lib.func("membrain_graph_load", voidPtr, ["str"]);
    const handle = fn_load(data);
    if (!handle) {
      const fn_err = lib.func("membrain_last_error", "str", []);
      throw new MembrainError(`failed to load graph: ${fn_err() ?? "unknown error"}`);
    }

    const instance = Object.create(MembrainGraph.prototype) as MembrainGraph;
    (instance as any).lib = lib;
    (instance as any).handle = handle;
    instance.bindGraphFunctions();
    return instance;
  }

  private bindGraphFunctions(): void {
    const lib = this.lib;

    this.fn_last_error = lib.func("membrain_last_error", "str", []);
    this.fn_string_free = lib.func("membrain_string_free", "void", [charPtr]);
    this.fn_graph_free = lib.func("membrain_graph_free", "void", [voidPtr]);

    this.fn_graph_add_node = lib.func("membrain_graph_add_node", "int32_t", [
      voidPtr, "str", "str", "double",
    ]);
    this.fn_graph_remove_node = lib.func("membrain_graph_remove_node", "int32_t", [
      voidPtr, "str",
    ]);
    this.fn_graph_query = lib.func("membrain_graph_query", "int32_t", [
      voidPtr, "str", "int32_t", "int32_t", charPtrPtr,
    ]);
    this.fn_graph_node_count = lib.func("membrain_graph_node_count", "int32_t", [
      voidPtr, int64Ptr,
    ]);
    this.fn_graph_edge_count = lib.func("membrain_graph_edge_count", "int32_t", [
      voidPtr, int64Ptr,
    ]);
    this.fn_graph_prune = lib.func("membrain_graph_prune", "int32_t", [
      voidPtr, charPtrPtr,
    ]);
    this.fn_graph_save = lib.func("membrain_graph_save", "int32_t", [
      voidPtr, charPtrPtr,
    ]);
    this.fn_graph_load = lib.func("membrain_graph_load", voidPtr, ["str"]);
  }

  private getLastError(): string {
    const err = this.fn_last_error();
    return err ?? "unknown error";
  }

  private check(code: number): void {
    if (code !== MEMBRAIN_OK) {
      throw new MembrainError(this.getLastError(), code);
    }
  }

  /** Add a node to the graph with an embedding vector. */
  addNode(memoryId: string, embedding: number[], confidence: number = 0.5): void {
    const embJson = JSON.stringify(embedding);
    const code = this.fn_graph_add_node(this.handle, memoryId, embJson, confidence);
    this.check(code);
  }

  /** Remove a node and all its incident edges. */
  removeNode(memoryId: string): void {
    const code = this.fn_graph_remove_node(this.handle, memoryId);
    this.check(code);
  }

  /** Run a multi-hop graph query using an embedding vector. */
  query(embedding: number[], maxHops: number = -1, topK: number = 10): GraphQueryResult {
    const embJson = JSON.stringify(embedding);
    const out = [null] as any;
    const code = this.fn_graph_query(this.handle, embJson, maxHops, topK, out);
    this.check(code);
    const json = out[0] as string;
    try {
      return JSON.parse(json) as GraphQueryResult;
    } finally {
      if (out[0]) this.fn_string_free(out[0]);
    }
  }

  /** Get the number of nodes in the graph. */
  nodeCount(): number {
    const out = newInt64Out();
    const code = this.fn_graph_node_count(this.handle, out);
    this.check(code);
    return readInt64(out);
  }

  /** Get the number of edges in the graph. */
  edgeCount(): number {
    const out = newInt64Out();
    const code = this.fn_graph_edge_count(this.handle, out);
    this.check(code);
    return readInt64(out);
  }

  /** Manually trigger graph pruning. */
  prune(): GraphPruningResult {
    const out = [null] as any;
    const code = this.fn_graph_prune(this.handle, out);
    this.check(code);
    const json = out[0] as string;
    try {
      return JSON.parse(json) as GraphPruningResult;
    } finally {
      if (out[0]) this.fn_string_free(out[0]);
    }
  }

  /** Save graph state to a base64-encoded string. */
  save(): string {
    const out = [null] as any;
    const code = this.fn_graph_save(this.handle, out);
    this.check(code);
    const data = out[0] as string;
    try {
      return data;
    } finally {
      if (out[0]) this.fn_string_free(out[0]);
    }
  }

  /** Destroy the graph and free all associated resources. */
  close(): void {
    if (this.handle) {
      this.fn_graph_free(this.handle);
      this.handle = null;
    }
  }
}

File: membrain-py/membrain/graph.py
"""Membrain graph memory layer client — backed by native PyO3 extension."""

from __future__ import annotations

import json
from typing import Any

from membrain._native import MembrainGraph as _NativeGraph

from .types import GraphPruningResult, GraphQueryResult, GraphScoredNode, GraphTraversalStep


class MembrainGraph:
    """Python client for the Membrain graph memory layer (true async via PyO3)."""

    def __init__(self, config: dict[str, Any] | None = None, *, lib_path: str | None = None) -> None:
        embedding_dim = config.get("embedding_dim") if config else None
        self._native = _NativeGraph(embedding_dim=embedding_dim)

    def _get_native(self) -> _NativeGraph:
        if not getattr(self, "_native", None):
            raise RuntimeError("graph is closed")
        return self._native

    def add_node(self, memory_id: str, embedding: list[float], confidence: float = 0.5) -> None:
        """Add a node to the graph with an embedding vector."""
        self._get_native().add_node(memory_id, embedding, confidence=confidence)

    def remove_node(self, memory_id: str) -> None:
        """Remove a node and all its incident edges."""
        self._get_native().remove_node(memory_id)

    async def query(self, embedding: list[float], max_hops: int = -1, top_k: int = 10) -> GraphQueryResult:
        """Run a multi-hop graph query using an embedding vector."""
        actual_hops = 2 if max_hops <= 0 else max_hops
        json_str = await self._get_native().query(embedding, max_hops=actual_hops, top_k=top_k)
        data = json.loads(json_str)
        return GraphQueryResult(
            nodes=[GraphScoredNode(**n) for n in data.get("nodes", [])],
            traversed_edges=[
                GraphTraversalStep(
                    from_id=s["from"],
                    to_id=s["to"],
                    edge_weight=s["edge_weight"],
                    attention_score=s["attention_score"],
                    hop=s["hop"],
                )
                for s in data.get("traversed_edges", [])
            ],
            hops_performed=data.get("hops_performed", 0),
            nodes_visited=data.get("nodes_visited", 0),
        )

    def node_count(self) -> int:
        """Get the number of nodes in the graph."""
        return self._get_native().node_count()

    def edge_count(self) -> int:
        """Get the number of edges in the graph."""
        return self._get_native().edge_count()

    def prune(self) -> GraphPruningResult:
        """Manually trigger graph pruning."""
        json_str = self._get_native().prune()
        data = json.loads(json_str)
        return GraphPruningResult(
            edges_removed=data.get("edges_removed", 0),
            nodes_removed=data.get("nodes_removed", 0),
            edges_remaining=data.get("edges_remaining", 0),
            nodes_remaining=data.get("nodes_remaining", 0),
        )

    def save(self) -> str:
        """Save graph state to a base64-encoded string."""
        return self._get_native().save()

    @classmethod
    def load(cls, data: str, *, lib_path: str | None = None) -> MembrainGraph:
        """Load graph state from a base64-encoded string."""
        instance = object.__new__(cls)
        instance._native = _NativeGraph.load(data, lib_path=lib_path)
        return instance

    def close(self) -> None:
        """Explicitly release the underlying native graph."""
        if getattr(self, "_native", None):
            self._native.close()
            self._native = None

    def __del__(self) -> None:
        """Clean up resources on deletion."""
        if hasattr(self, 'close'):
            self.close()

    def __enter__(self) -> MembrainGraph:
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()

    def __repr__(self) -> str:
        return repr(self._native) if hasattr(self, "_native") and self._native is not None else "MembrainGraph(closed)"
